bftengine/src/bcstatetransfer/BCStateTran.hpp

```diff
@@ -357,21 +357,80 @@ class BCStateTran : public IStateTransfer {
   ///////////////////////////////////////////////////////////////////////////
   // Asynchronous Operations - Blocks IO
   ///////////////////////////////////////////////////////////////////////////
-  struct GetBlockContext {
-    uint16_t index;
-    uint64_t blockId;
-    uint32_t blockSize;
-    std::unique_ptr<char[]> block;
+
+  struct BlockIOContext {
+    uint64_t blockId = 0;
+    uint32_t actualBlockSize = 0;
+    std::unique_ptr<char[]> blockData;
     std::future<bool> future;
   };
 
-  std::vector<GetBlockContext> srcGetBlockContextes_;
+  using BlockIOContextPtr = std::shared_ptr<BlockIOContext>;
+  class BlockIODataPool {
+   public:
+    BlockIODataPool(size_t maxNumElements, size_t elementSize, FetchingState& fetchingState)
+        : maxNumElements_(maxNumElements), fetchingState_(fetchingState) {
+      for (size_t i{0}; i < maxNumElements_; ++i) {
+        auto element = std::make_shared<BlockIOContext>();
+        element->blockData.reset(new char[elementSize]);
+        freeQ_.push_back(element);
+      }
+    }
+
+    size_t numFreeElements() const { return freeQ_.size(); }
+    size_t numAllocatedElements() const { return allocatedQ_.size(); }
+    bool empty() { return freeQ_.empty(); };
+    bool full() { return allocatedQ_.empty(); };
+    size_t maxElements() { return maxNumElements_; };
+
+    BlockIOContextPtr alloc() {
+      if (freeQ_.empty()) {
+        throw std::runtime_error("No more free elements!");
```

Comment:
what happens in this case? do we want to fail the entire replica?

Reply:
This should never happen, it means I have a bug in my code. The caller should be responsible to make sure there are always enough elements. It is doing so by calling processCommitResultsAsync with the right flags.
if there are no more free elements in the pool, it should call processCommitResultsAsync with breakIfNotReady=false.
In practice, it will never happen anyway. We have no more than 5-6 concurrent at peak, while the batch size (used also in the source) is much larger.